Sources/ImgixSwift/ImgixClient.swift

```diff
@@ -98,6 +98,7 @@ import Foundation
         }
         else {
             path = path.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? path
```

Comment:
Would something like

```swift
path.addingPercentEncoding(
    withAllowedCharacters: CharacterSet(charactersIn: "#?:+").inverted
)
```

work here?

Reply:
Interesting, but why not use the builtin `withAllowedCharacters`? They [adhere to the RFC3986 standard](https://stackoverflow.com/a/44755781/11760796), which happens to include all the chars we want to encode.

Reply:
Using `.urlPathAllowed` is part of the issue. Say we want to exclude `+`, `:`, `#`, and `?` from the allowed character set. We could do

```swift
path = path.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? path
path = path.replacingOccurrences(of: "+", with: "%2B") // Assuming this iterates over the entire string, again.
// this doesn't handle `:` either...
```

but it would be nice to do the encoding we want in one pass (i.e. one call to ` path.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed)` and done).

```swift
urlPathAllowed.contains("#") // false, good
urlPathAllowed.contains("?") // false, good
urlPathAllowed.contains(":") // true
urlPathAllowed.contains("+") // true
```

So another solution would be to drop these characters from the allowed set, from `urlPathAllowed`: 
```swift
var allowedChars = CharacterSet.urlPathAllowed
allowedChars.remove("+")
allowedChars.remove(":")
path = path.addingPercentEncoding(withAllowedCharacters: allowedChars) ?? path
```

And this should result in the behavior we are looking for:

```swift
func testEncodesPlusHashColonAndQuestionMarkChars() {
    let generatedUrl = client.buildUrl(":+?#.jpg", params: [:])
    let expectedUrl = "https://paulstraw.imgix.net/%3A%2B%3F%23.jpg"
    XCTAssert(generatedUrl.absoluteString == expectedUrl)
}
```

Reply:
@ericdeansanchez you were 💯  right that this could be done. We'd actually done it before 😆 . Take a look at [ixEncodeURIComponent](https://github.com/imgix/imgix-swift/blob/15d17086d188300c6680d9a746ed23fccbfcec7d/Sources/ImgixSwift/String%2BImgixSwift.swift#L11-L16). We do exactly the inverse of this. From there it wasn't hard to think up a solution:

```swift
    static var ixEncodeUriCharSet: CharacterSet = {
        var charSet = CharacterSet.urlPathAllowed
        charSet.remove(charactersIn: "#?:+")
        return charSet
    }()
...
func ixEncodeUri() -> String {
        return self.addingPercentEncoding(withAllowedCharacters: String.ixEncodeUriCharSet)!
    }
```